crates/polars-core/src/series/arithmetic/borrowed.rs

```diff
@@ -171,6 +175,168 @@ impl NumOpsDispatchInner for FixedSizeListType {
     }
 }
 
+/// Given an ArrayRef with some primitive values, wrap it in list(s) until it
+/// matches the requested shape.
+fn reshape_list_based_on(data: &ArrayRef, shape: &ArrayRef) -> PolarsResult<ArrayRef> {
+    if let Some(list_chunk) = shape.as_any().downcast_ref::<LargeListArray>() {
+        let result = LargeListArray::try_new(
+            list_chunk.dtype().clone(),
+            list_chunk.offsets().clone(),
+            reshape_list_based_on(data, list_chunk.values())?,
+            list_chunk.validity().cloned(),
+        )?;
+        Ok(Box::new(result))
+    } else {
+        Ok(data.clone())
+    }
+}
+
+/// Given an ArrayRef, return true if it's a LargeListArrays and it has one or
+/// more nulls.
+fn maybe_list_has_nulls(data: &ArrayRef) -> bool {
+    if let Some(list_chunk) = data.as_any().downcast_ref::<LargeListArray>() {
+        if list_chunk
+            .validity()
+            .map(|bitmap| bitmap.unset_bits() > 0)
+            .unwrap_or(false)
+        {
+            true
+        } else {
+            maybe_list_has_nulls(list_chunk.values())
+        }
+    } else {
+        false
+    }
+}
+
+/// Return whether the left and right have the same shape. We assume neither has
+/// any nulls, recursively.
+fn lists_same_shapes(left: &ArrayRef, right: &ArrayRef) -> bool {
+    let left_as_list = left.as_any().downcast_ref::<LargeListArray>();
+    let right_as_list = right.as_any().downcast_ref::<LargeListArray>();
+    match (left_as_list, right_as_list) {
+        (Some(left), Some(right)) => {
+            left.offsets() == right.offsets() && lists_same_shapes(left.values(), right.values())
+        },
+        (None, None) => left.len() == right.len(),
+        _ => false,
+    }
+}
+
+impl ListChunked {
+    fn arithm_helper(
+        &self,
+        rhs: &Series,
+        op: &dyn Fn(&Series, &Series) -> PolarsResult<Series>,
+    ) -> PolarsResult<Series> {
+        polars_ensure!(
+            self.len() == rhs.len(),
+            InvalidOperation: "can only do arithmetic operations on Series of the same size; got {} and {}",
+            self.len(),
+            rhs.len()
+        );
+
+        thread_local! {
```

Comment:
Why do we deal with `thread_local` and statics here?

Reply:
If this makes sense I'll add an explanatory comment.

Reply:
And I should add, recursive in a way that doesn't allow passing arguments, since there's a whole bunch of layers in between.

Reply:
We should pass the the arguments. Going thread_local isn't an option I want for arithmetic, and to be fair almost never in the codebase.. I also think it is incorrect. A thread can be halted in the middle of this work and then another rayon task can start on this thread and overwrite the thread local state.

Reply:
This would involve modifying a bunch of more general traits, e.g. `PrivateSeries::add_to`, to have an extra argument. Which would be somewhat intrusive, but possible I guess.

Another option is just to stick to the slow path only.

Reply:
A third option is that I'm wrong and recursion isn't necessary, or could be done in a way that bypasses the more general-purpose APIs. So I'll try to see if I can figure that out, and report back.

Reply:
OK, figured it out.